specs/gloas/beacon-chain.md

```diff
@@ -1048,6 +1190,197 @@ def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
+            withdrawal_request.amount,
+        )
+        exit_queue_epoch = compute_exit_epoch_and_update_churn(state, to_withdraw)
+        withdrawable_epoch = Epoch(exit_queue_epoch + MIN_VALIDATOR_WITHDRAWABILITY_DELAY)
+        state.pending_partial_withdrawals.append(
+            PendingPartialWithdrawal(
+                validator_index=validator_index,
+                amount=to_withdraw,
+                withdrawable_epoch=withdrawable_epoch,
+            )
+        )
+```
+
+###### Modified `process_withdrawal_request`
+
+```python
+def process_withdrawal_request(state: BeaconState, withdrawal_request: WithdrawalRequest) -> None:
+    # If partial withdrawal queue is full, only full exits are processed
+    is_queue_full = len(state.pending_partial_withdrawals) == PENDING_PARTIAL_WITHDRAWALS_LIMIT
+    is_full_exit_request = withdrawal_request.amount == FULL_EXIT_REQUEST_AMOUNT
+    if is_queue_full and not is_full_exit_request:
+        return
+
+    if is_builder(state, withdrawal_request.validator_pubkey):
+        builder_index = get_builder_index(state, withdrawal_request.validator_pubkey)
+        return process_withdrawal_request_for_builder(state, withdrawal_request, builder_index)
+    elif is_validator(state, withdrawal_request.validator_pubkey):
+        validator_index = get_validator_index(state, withdrawal_request.validator_pubkey)
+        return process_withdrawal_request_for_validator(state, withdrawal_request, validator_index)
+```
+
+##### Deposit requests
+
+###### New `get_index_for_new_builder`
+
+```python
+def get_index_for_new_builder(state: BeaconState) -> BuilderIndex:
+    for index, builder in enumerate(state.builders):
+        if builder.withdrawable_epoch <= get_current_epoch(state) and builder.balance == 0:
+            return BuilderIndex(index)
+    return BuilderIndex(len(state.builders))
+```
+
+###### New `get_builder_from_deposit`
+
+```python
+def get_builder_from_deposit(
+    state: BeaconState, pubkey: BLSPubkey, withdrawal_credentials: Bytes32, amount: uint64
+) -> Builder:
+    return Builder(
+        pubkey=pubkey,
+        withdrawal_credentials=withdrawal_credentials,
+        balance=amount,
+        deposit_epoch=get_current_epoch(state),
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+    )
+```
+
+###### New `add_validator_to_registry`
+
+```python
+def add_builder_to_registry(
+    state: BeaconState, pubkey: BLSPubkey, withdrawal_credentials: Bytes32, amount: uint64
+) -> None:
+    index = get_index_for_new_builder(state)
+    builder = get_builder_from_deposit(state, pubkey, withdrawal_credentials, amount)
+    set_or_append_list(state.builders, index, builder)
+```
+
+###### New `apply_deposit_for_builder`
+
+```python
+def apply_deposit_for_builder(
+    state: BeaconState,
+    pubkey: BLSPubkey,
+    withdrawal_credentials: Bytes32,
+    amount: uint64,
+    signature: BLSSignature,
+) -> None:
+    if is_validator(state, pubkey):
```

Comment:
Regardless of withdrawal credentials type if a validator with suck pubkey exists, I think it should be processed as a top-up